pkg/controllers/status/controller.go

```diff
@@ -68,181 +73,148 @@ const (
-		metrics:                       controllerConfig.Metrics,
-		logger:                        logger,
-		eventRecorder:                 eventsink.NewDefederatingRecorderMux(kubeClient, StatusControllerName, 6),
+		fedClient:                      fedClient,
+		fedInformerManager:             fedInformerManager,
+		ftcManager:                     ftcManager,
+		fedObjectInformer:              fedObjectInformer,
+		clusterFedObjectInformer:       clusterFedObjectInformer,
+		collectedStatusInformer:        collectedStatusInformer,
+		clusterCollectedStatusInformer: clusterCollectedStatusInformer,
+		clusterAvailableDelay:          clusterAvailableDelay,
+		clusterUnavailableDelay:        clusterUnavailableDelay,
+		reconcileOnClusterChangeDelay:  time.Second * 3,
+		memberObjectEnqueueDelay:       memberObjectEnqueueDelay,
+		metrics:                        metrics,
+		logger:                         logger.WithValues("controller", StatusControllerName),
+		eventRecorder:                  eventsink.NewDefederatingRecorderMux(kubeClient, StatusControllerName, 4),
 	}
 
 	s.worker = worker.NewReconcileWorker(
+		StatusControllerName,
+		nil,
 		s.reconcile,
 		worker.RateLimiterOptions{},
-		controllerConfig.WorkerCount,
-		controllerConfig.Metrics,
-		delayingdeliver.NewMetricTags("status-worker", typeConfig.GetTargetType().Kind),
+		workerCount,
+		metrics,
 	)
 
 	// Build queue for triggering cluster reconciliations.
 	s.clusterQueue = workqueue.NewNamedDelayingQueue("status-controller-cluster-queue")
 
-	// Start informers on the resources for the federated type
-	enqueueObj := s.worker.EnqueueObject
+	handler := util.NewTriggerOnAllChanges(func(o pkgruntime.Object) {
+		s.worker.Enqueue(common.NewQualifiedName(o))
+	})
 
-	targetNamespace := controllerConfig.TargetNamespace
+	s.fedObjectInformer.Informer().AddEventHandler(handler)
+	s.clusterFedObjectInformer.Informer().AddEventHandler(handler)
 
-	s.federatedStore, s.federatedController = util.NewResourceInformer(
-		federatedTypeClient,
-		targetNamespace,
-		enqueueObj,
-		controllerConfig.Metrics,
-	)
-	s.statusStore, s.statusController = util.NewResourceInformer(
-		statusClient,
-		targetNamespace,
-		enqueueObj,
-		controllerConfig.Metrics,
-	)
-	logger.Info("Creating new FederatedInformer")
-
-	targetAPIResource := typeConfig.GetTargetType()
-
-	// Federated informer for resources in member clusters
-	s.informer, err = util.NewFederatedInformer(
-		controllerConfig,
-		client,
-		configCopy,
-		&targetAPIResource,
-		func(obj pkgruntime.Object) {
-			qualifiedName := common.NewQualifiedName(obj)
-			s.worker.EnqueueWithDelay(qualifiedName, s.memberObjectEnqueueDelay)
+	s.collectedStatusInformer.Informer().AddEventHandler(handler)
+	s.clusterCollectedStatusInformer.Informer().AddEventHandler(handler)
+
+	if err := s.fedInformerManager.AddEventHandlerGenerator(&informermanager.EventHandlerGenerator{
+		Predicate: informermanager.RegisterOncePredicate,
```

Comment:
Since status collection can be enabled/disabled in the FTC, I think we need to check for this in the predicate.

Reply:
Realized it doesn't affect correctness since we check for status collection enablement during reconcile anyway. But it would be good to reduce unnecessary processing as well.